<a href="https://colab.research.google.com/github/Oluwabukunmige/My-Machine-Learning-Journey/blob/master/Regression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [10]:
!pip install PyDrive

In [0]:
import os
from pydrive.auth import GoogleAuth
from pydrive.drive import GoogleDrive
from google.colab import auth
from oauth2client.client import GoogleCredentials

In [0]:
auth.authenticate_user()
gauth = GoogleAuth()
gauth.credentials = GoogleCredentials.get_application_default()
drive = GoogleDrive(gauth)

In [0]:
download = drive.CreateFile({'id': '1KpWx-lPL6-zNJpz_GvaibI3CqqVhuDtn'})
download.GetContentFile('cal_housing.csv')

In [0]:
import pandas as pd
cal_housing = pd.read_csv('cal_housing.csv')


In [0]:
final_housing = cal_housing.copy()

In [16]:
final_housing.head()

longitude  latitude  ...  median_house_value  ocean_proximity
0    -122.23     37.88  ...            452600.0         NEAR BAY
1    -122.22     37.86  ...            358500.0         NEAR BAY
2    -122.24     37.85  ...            352100.0         NEAR BAY
3    -122.25     37.85  ...            341300.0         NEAR BAY
4    -122.25     37.85  ...            342200.0         NEAR BAY

[5 rows x 10 columns]

In [9]:
final_housing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 10 columns):
longitude             20640 non-null float64
latitude              20640 non-null float64
housing_median_age    20640 non-null float64
total_rooms           20640 non-null float64
total_bedrooms        20433 non-null float64
population            20640 non-null float64
households            20640 non-null float64
median_income         20640 non-null float64
median_house_value    20640 non-null float64
ocean_proximity       20640 non-null object
dtypes: float64(9), object(1)
memory usage: 1.6+ MB


The total_bedrooms column has some missing values. 

In [17]:
cal_housing['total_bedrooms'].value_counts(dropna = False)


NaN       207
280.0      55
331.0      51
345.0      50
393.0      49
343.0      49
394.0      48
328.0      48
348.0      48
272.0      47
309.0      47
295.0      46
314.0      46
322.0      46
399.0      46
317.0      46
284.0      45
388.0      45
290.0      45
291.0      45
346.0      45
287.0      45
340.0      45
313.0      45
269.0      44
460.0      44
365.0      44
294.0      44
361.0      44
312.0      44
         ... 
2814.0      1
6445.0      1
2141.0      1
2460.0      1
3479.0      1
1758.0      1
1127.0      1
2010.0      1
1571.0      1
3224.0      1
1494.0      1
1052.0      1
2574.0      1
1570.0      1
1437.0      1
2118.0      1
3336.0      1
2289.0      1
1736.0      1
940.0       1
980.0       1
1215.0      1
2009.0      1
3864.0      1
1288.0      1
1691.0      1
2205.0      1
1448.0      1
1887.0      1
2546.0      1
Name: total_bedrooms, Length: 1924, dtype: int64

There are 207 missing values which is quite significant. I will fill the missing values using the median value in the total_beadroom column

In [0]:
final_housing['total_bedrooms'].fillna(final_housing['total_bedrooms'].median(), inplace = True)

In [0]:
final_housing['total_bedrooms'].value_counts()

Preprocess the columns

In [0]:
final_housing['longitude'].value_counts()

In [0]:
final_housing['latitude'].describe()

In [0]:
final_housing['housing_median_age'].describe()

In [0]:
final_housing.drop('ocean_proximity', axis = 1, inplace = True)

In [20]:
final_housing['median_house_value'].value_counts()

500001.0    965
137500.0    122
162500.0    117
112500.0    103
187500.0     93
225000.0     92
350000.0     79
87500.0      78
275000.0     65
150000.0     64
175000.0     63
100000.0     62
125000.0     56
67500.0      55
250000.0     47
200000.0     46
118800.0     39
450000.0     37
156300.0     35
212500.0     33
193800.0     31
181300.0     31
300000.0     30
75000.0      30
81300.0      29
55000.0      29
93800.0      29
106300.0     28
143800.0     28
400000.0     27
           ... 
373500.0      1
433000.0      1
354600.0      1
389700.0      1
370100.0      1
304200.0      1
477200.0      1
422500.0      1
445900.0      1
309000.0      1
413100.0      1
465700.0      1
353700.0      1
438900.0      1
396700.0      1
380300.0      1
478500.0      1
455300.0      1
403000.0      1
363900.0      1
494400.0      1
389800.0      1
347500.0      1
487000.0      1
331100.0      1
303200.0      1
307900.0      1
383200.0      1
360800.0      1
405500.0      1
Name: median_house_value

In [0]:
X = final_housing.drop('median_house_value', axis = 1)
y = final_housing['median_house_value']

In [0]:
X.head()

In [0]:
X.info()


Import the needed modules

In [0]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import Ridge
from sklearn.linear_model import Lasso

In [0]:
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 0) 


Scale the data

In [0]:
scaler = StandardScaler()
XTrain_scaled = scaler.fit_transform(X_train)
XTest_scaled = scaler.transform(X_test)

Implement Linear Regression

In [28]:
lr =  LinearRegression()
lr.fit(XTrain_scaled,y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [29]:
lr.score(XTrain_scaled, y_train)

0.6386174149948768

In [30]:
lr.score(XTest_scaled, y_test)

0.625834759675782

In [0]:
ri = Ridge()

In [0]:
ri.fit(XTrain_scaled,y_train)

Ridge(alpha=1.0, copy_X=True, fit_intercept=True, max_iter=None,
      normalize=False, random_state=None, solver='auto', tol=0.001)

In [0]:
ri.score(XTrain_scaled,y_train)

0.6386173352454366

In [0]:
ri.score(XTest_scaled,y_test)

0.6258276733020272

In [0]:
la = Lasso()
la.fit(XTrain_scaled, y_train)

Lasso(alpha=1.0, copy_X=True, fit_intercept=True, max_iter=1000,
      normalize=False, positive=False, precompute=False, random_state=None,
      selection='cyclic', tol=0.0001, warm_start=False)

In [0]:
la.score(XTrain_scaled, y_train)

0.6386174095728411

In [0]:
la.score(XTest_scaled, y_test)

0.6258325066181337

In [0]:
polynomial_features= PolynomialFeatures(degree = 3)
x_train = polynomial_features.fit_transform(XTrain_scaled)
x_test = polynomial_features.transform(XTest_scaled)

In [37]:
lr.fit(x_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [38]:
lr.score(x_train, y_train)

0.7407055037928293

In [39]:
lr.score(x_test, y_test)

0.6761811021164867

In [40]:
ri.fit(x_train, y_train)

Ridge(alpha=1.0, copy_X=True, fit_intercept=True, max_iter=None,
      normalize=False, random_state=None, solver='auto', tol=0.001)

In [41]:
ri.score(x_train, y_train)

0.7405583493530259

In [42]:
ri.score(x_test, y_test)

0.6763783324100713